pkg/sorter/sorter.go

```diff
@@ -0,0 +1,111 @@
+	// If it's based on table, fetching events by table is prefered.
+	IsTableBased() bool
+
+	// AddTable adds the table into the engine.
+	AddTable(tableID model.TableID)
+
+	// RemoveTable removes the table from the engine.
+	RemoveTable(tableID model.TableID)
+
+	// Add adds the given events into the sort engine.
+	//
+	// NOTE: it's an asynchronous interface. To get the notification of when
+	// events are available for fetching, OnResolve is what you want.
+	Add(tableID model.TableID, events ...*model.PolymorphicEvent) error
+
+	// OnResolve pushes action into EventSortEngine's hook list, which
+	// will be called after any events are resolved.
+	OnResolve(action func(model.TableID, model.Ts))
+
+	// FetchByTable creates an iterator to fetch events from the given table.
+	// lowerBound is inclusive and only resolved events can be retrieved.
+	//
+	// NOTE: FetchByTable is always available even if IsTableBased returns false.
+	FetchByTable(tableID model.TableID, lowerBound, upperBound Position) EventIterator
+
+	// FetchAllTables creates an iterator to fetch events from all tables.
+	// lowerBound is inclusive and only resolved events can be retrieved.
+	//
+	// NOTE: It's only available if IsTableBased returns true.
+	FetchAllTables(lowerBound Position) EventIterator
+
+	// CleanByTable tells the engine events of the given table in the given range
+	// (unlimited, upperBound] are committed and not necessary any more.
+	// The EventSortEngine instance can GC them later.
+	//
+	// NOTE: CleanByTable is always available even if IsTableBased returns false.
+	CleanByTable(tableID model.TableID, upperBound Position) error
+
+	// CleanAllTables tells the engine events of all tables in the given range
+	// (unlimited, upperBound] are committed and not necessary any more.
+	// The EventSortEngine instance can GC them later.
+	//
+	// NOTE: It's only available if IsTableBased returns true.
+	CleanAllTables(upperBound Position) error
+
+	// Close closes the engine. All data written by this instance can be deleted.
+	//
+	// NOTE: it leads an undefined behavior to close an engine with active iterators.
+	Close() error
+}
+
+// EventIterator is an iterator to fetch events from EventSortEngine.
+// It's unnecessary to be thread-safe.
+type EventIterator interface {
+	// Next is used to fetch one event. nil indicates it reaches the stop point.
+	//
+	// txnFinished indicates whether all events in the current transaction are
+	// fetched or not. Users should keep fetching events until txnFinished.Valid()
+	// returns true.
+	//
+	// NOTE: event.IsResolved() will always be false.
+	Next() (event *model.PolymorphicEvent, txnFinished Position, err error)
+
+	// Close closes the iterator.
+	Close() error
+}
+
+// Position is used to
+//  1. fetch or clear events from an engine, for example, see EventSortEngine.FetchByTable.
+//  2. calculate the next position with method Next.
+type Position struct {
```

Comment:
Could you please add a `Less` function to it?

Reply:
I prefer `Compare` instead of `Less`.